client/protocols/testing.py

```diff
@@ -0,0 +1,287 @@
+#    Testing Mechanism    #
+###########################
+
+# Must get current dir for Travis to pass, among other reasons
+CURR_DIR = os.getcwd()
+# Users will generally name their test file the following name. 
+# If changing default name, change in ok.py args parse as well
+DEFAULT_TST_FILE = "mytests.rst"
+
+def timeout(seconds_before_timeout):
+    """www.saltycrane.com/blog/2010/04/using-python-timeout-decorator-uploading-s3/"""
+    def decorate(f):
+        def handler(signum, frame):
+            raise EarlyExit("Test examples timed out!")
+        def new_f(*args, **kwargs):
+            old = signal.signal(signal.SIGALRM, handler)
+            signal.alarm(seconds_before_timeout)
+            try:
+                result = f(*args, **kwargs)
+            finally:
+                signal.signal(signal.SIGALRM, old)
+            signal.alarm(0)
+            return result
+        new_f.__name__ = f.__name__
+        return new_f
+    return decorate
+
+class TestingProtocol(models.Protocol):
+    """A Protocol that executes doctests as lists of Example objects, supports 
+    suite/case specificity, alternate file testing, and provides users with 
+    details such as cases passed and test coverage.
+    """
+    def __init__(self, args, assignment):
+        super().__init__(args, assignment)
+        # The environment in which the doctests are run (global vars)
+        self.good_env = {}
+        self.verb = self.args.verbose
+        # Initialize the doctest module objects that will do the testing/parse
+        self.parser = DocTestParser()
+        self.runner = DocTestRunner(verbose=self.verb, optionflags=FAIL_FAST)
+        self.lines_exec = 0
+        self.lines_total = 0
+
+
+    def test(self, good_env={}, suite=None, case=None):
+        test_results = {}
+        # all examples to be run will be put in exs
+        exs = collections.OrderedDict()
+        # use regex to get raw strings organized into suite/case
+        self.get_data()
+        try:
+            if suite:
+                exs = self.get_suite_examples(suite, case)
+            elif case:
+                # No support for cases without their suite
+                raise EarlyExit('python3 ok: error: ' 
+                    'Please specify suite for given case ({}).'.format(case[0]))
+            else:
+                exs = self.get_all_examples()
+            # gets analytics to be returned
+            test_results[self.tstfile_name] =  self.analyze(suite, case, exs)
+        except KeyError as e:
+            raise EarlyExit('python3 ok: error: ' 
+                    'Suite/Case label must be valid.'
+                    '(Suites: {}, Cases: {})'.format(self.num_suites, self.num_cases))
+        return test_results
+
+    def analyze(self, suite, case, examples):
+        rfailed, rattempted = self.run_examples(examples)
+        self.postcov.stop()
+        failed = rfailed
+        attempted = rattempted
+        passed = attempted - failed
```

Comment:
Is there a reason why `failed` and `attempted` variables are created instead of just using `failed` and `rattempted`?

Reply:
Ah I didn't catch that!